pkg/controllers/disruption/helpers.go

```diff
@@ -282,6 +225,67 @@ func BuildNodePoolMap(ctx context.Context, kubeClient client.Client, cloudProvid
 	return nodePoolMap, nodePoolToInstanceTypesMap, nil
 }
 
+// BuildDisruptionBudgets prepares our disruption budget mapping. The disruption budget maps each disruption reason to the number of allowed disruptions.
+// We calculate allowed disruptions by taking the max disruptions allowed by disruption reason and subtracting the number of nodes that are NotReady and already being deleted by that disruption reason.
+//
+//nolint:gocyclo
+func BuildDisruptionBudgets(ctx context.Context, cluster *state.Cluster, clk clock.Clock, kubeClient client.Client, recorder events.Recorder) (map[string]map[v1beta1.DisruptionReason]int, error) {
+	disruptionBudgetMapping := map[string]map[v1beta1.DisruptionReason]int{}
+	numNodes := map[string]int{}   // map[nodepool] -> node count in nodepool
+	disrupting := map[string]int{} // map[nodepool] -> nodes undergoing disruption
+	for _, node := range cluster.Nodes() {
+		// We only consider nodes that we own and are initialized towards the total.
+		// If a node is launched/registered, but not initialized, pods aren't scheduled
+		// to the node, and these are treated as unhealthy until they're cleaned up.
+		// This prevents odd roundup cases with percentages where replacement nodes that
+		// aren't initialized could be counted towards the total, resulting in more disruptions
+		// to active nodes than desired, where Karpenter should wait for these nodes to be
+		// healthy before continuing.
+		if !node.Managed() || !node.Initialized() {
+			continue
+		}
+
+		nodePool := node.Labels()[v1beta1.NodePoolLabelKey]
+		numNodes[nodePool]++
+
+		// If the node satisfies one of the following, we subtract it from the allowed disruptions.
+		// 1. Has a NotReady conditiion
+		// 2. Is marked as disrupting
+		if cond := nodeutils.GetCondition(node.Node, v1.NodeReady); cond.Status != v1.ConditionTrue || node.MarkedForDeletion() {
+			disrupting[nodePool]++
+		}
+	}
```

Comment:
We could break this down into a helper function for calculating numNodes and disrupting but I would prefer we don't obscure the code here and just ignore the nocyclo. 

Its clearer to me. Thoughts?